cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaAssemblyOperator.java

```diff
@@ -83,40 +87,151 @@ public KafkaAssemblyOperator(Vertx vertx, boolean isOpenShift,
+        private final Secret clientsCASecret;
+        private final Secret clientsPublicKeySecret;
+        private final Secret brokersClientsSecret;
+        private final Secret brokersInternalSecret;
+        private ReconcileResult<StatefulSet> diffs;
+
+        KafkaClusterDescription(KafkaCluster kafka, Service service, Service headlessService,
+                                ConfigMap metricsConfigMap, StatefulSet statefulSet,
+                                Secret clientsCASecret, Secret clientsPublicKeySecret,
+                                Secret brokersClientsSecret, Secret brokersInternalSecret) {
+            this.kafka = kafka;
+            this.service = service;
+            this.headlessService = headlessService;
+            this.metricsConfigMap = metricsConfigMap;
+            this.statefulSet = statefulSet;
+            this.clientsCASecret = clientsCASecret;
+            this.clientsPublicKeySecret = clientsPublicKeySecret;
+            this.brokersClientsSecret = brokersClientsSecret;
+            this.brokersInternalSecret = brokersInternalSecret;
+        }
+
+        KafkaCluster kafka() {
+            return this.kafka;
+        }
+
+        Service service() {
+            return this.service;
+        }
+
+        Service headlessService() {
+            return this.headlessService;
+        }
+
+        ConfigMap metricsConfigMap() {
+            return this.metricsConfigMap;
+        }
+
+        StatefulSet statefulSet() {
+            return this.statefulSet;
+        }
+
+        Secret clientsCASecret() {
+            return this.clientsCASecret;
+        }
+
+        Secret clientsPublicKeySecret() {
+            return this.clientsPublicKeySecret;
+        }
+
+        Secret brokersClientsSecret() {
+            return this.brokersClientsSecret;
+        }
+
+        Secret brokersInternalSecret() {
+            return this.brokersInternalSecret;
+        }
+
+        ReconcileResult<StatefulSet> diffs() {
+            return this.diffs;
+        }
+
+        Future<KafkaClusterDescription> withDiff(Future<ReconcileResult<StatefulSet>> r) {
+            return r.map(rr -> {
+               this.diffs = rr;
+               return this;
+            });
+        }
+
+        Future<KafkaClusterDescription> withVoid(Future<?> r) {
+            return r.map(this);
+        }
+    }
+
+    private final Future<KafkaClusterDescription> getKafkaCluster(ConfigMap assemblyCm, List<Secret> assemblySecrets) {
+        Future<KafkaClusterDescription> fut = Future.future();
+
+        vertx.createSharedWorkerExecutor("kubernetes-ops-pool").executeBlocking(
+                future -> {
+                    try {
+                        KafkaCluster kafka = KafkaCluster.fromDescription(assemblyCm, assemblySecrets);
+
+                        KafkaClusterDescription desc =
+                                new KafkaClusterDescription(kafka, kafka.generateService(), kafka.generateHeadlessService(),
```

Comment:
Why exactly do we need the `KafkaClusterDescription` class which just wraps values from `KafkaCluster`? In what way is this better than returning `KafkaCluster`?

Reply:
The idea is that such a class is useful to bring result from a lambda to another in the composite chain. It's true that the only main import value is the `KafkaCluster` one but we also need Services, ConfigMaps and so on. We could think to remove them calling every time the `kafka.generateXXXX` method where needed but the description class is still needed because at some point the lambda result needs to be enriched by the Diffs result (used by the next one).
My first implementation (due to effectively final problem) was to declare the `KafkaCluster' instance as a class member making the assembly operator more stateful so @tombentley suggested this more elegant way.